### DISCUSSION TITLE
How can I provide "--compile --install-option='--with-openssl' pycurl" these options while adding pycurl dependency?

### DISCUSSION THREAD

**Discussion by paraggit** [python-poetry/poetry#6814] [General] [answered]
My Project required pycurl as dependency and i have added it via 'poetry add pycurl'.  Its get added but while importing module its showing following error.  And to resolve this i need to add "--compile --install-option='--with-openssl'" these options while adding dependency.  Is there any way to add those options? and it should support in 'poetry install' command also. Thanks. 
```
In [4]: import pycurl
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-4-141165d68a5f> in <module>
----> 1 import pycurl

ImportError: pycurl: libcurl link-time ssl backends (secure-transport, openssl) do not include compile-time ssl backend (none/other)
```

**Comment by clintonroy** [comment, marked as answer]
This is not possible currently, and I do not see such a feature ever being on the roadmap.

The python ecosystem is very much moving towards installing things from wheels, not from source. 

If you really need something from the curl package, I would build it into a wheel and use that. I would then push those patches upstream.

Otherwise, I'd strongly recommend that you look at using packages that provide similar functionality, and are packaged as wheels, requests, for example.

**Reply by paraggit** [reply]
Thanks @clintonroy for your inputs.

**Comment by deeenes** [comment]
I don't think this is a good or satisfying solution. You suggest using another package only because of this issue -- indeed requests or urllib3 are alternatives of pycurl for maybe 90% of the use cases, but they can not do everything that pycurl can; there is a reason why pycurl exists. And this approach also suggests that installing a package linked against two pretty standard C libraries (libcurl & openssl) is an impossible task for pip and poetry. Which is not true, we install these kind of packages all the time.

Pycurl is not distributed as wheel in PyPI, and I think for a good reason: to force pip to compile it locally, hence to link it against the library versions available on the target system. If it was distributed as a wheel, it would very often throw an import error like `libcurl link-time version is older than compile-time version`. It's possible to include dylibs in wheels, but it [requires a special tool](https://github.com/matthew-brett/delocate), and because of that it feels a hacky workaround. Then it would be even better to compile statically.

If pycurl is compiled by pip on a correctly configured system it should work just fine. I think `with-openssl` is the default for pycurl. If it does not compile with OpenSSL that means it is not found in the include and library paths. Hence it is a broader issue with the general configuration of the system environment and the installation and management of various packages. This issue has nothing to do with pycurl or poetry, but rather with conda, homebrew and OSX. To address these configuration issues, the `LDFLAGS` and `CPPFLAGS` can be used; also `with-openssl` can be set by an env variable (of course set them to the actual paths you have your openssl in, alternatively, you can also create the symlinks how Brew does it for non keg-only packages):

```
PYCURL_SSL_LIBRARY=openssl LDFLAGS="-L/usr/local/opt/openssl/lib" CPPFLAGS="-I/usr/local/opt/openssl/include" pip install --no-cache-dir --force-reinstall pycurl
```

Some packages such as `lxml` [go a long way](https://github.com/lxml/lxml/blob/master/buildlibxml.py) to work around this issue: this custom script downloads the sources of four libraries, takes care of the configuration and compiles them by `configure` and `make`. A similar script could be added to the pycurl package. Apparently this script is called by `setup.py` and links the libraries statically if building a wheel for distribution, but links them dynamically if it's compiled locally.
